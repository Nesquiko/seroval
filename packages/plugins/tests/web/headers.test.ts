import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import HeadersPlugin from '../../web/headers';

const EXAMPLE = new Headers([
  ['Content-Type', 'text/plain'],
  ['Content-Encoding', 'gzip'],
]);

describe('Headers', () => {
  describe('serialize', () => {
    it('supports Headers', () => {
      const result = serialize(EXAMPLE, {
        plugins: [HeadersPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back).toBeInstanceOf(Headers);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('serializeAsync', () => {
    it('supports Headers', async () => {
      const result = await serializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [HeadersPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof EXAMPLE>>(result);
      expect(back).toBeInstanceOf(Headers);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('toJSON', () => {
    it('supports Headers', () => {
      const result = toJSON(EXAMPLE, {
        plugins: [HeadersPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Headers>(result, {
        plugins: [HeadersPlugin],
      });
      expect(back).toBeInstanceOf(Headers);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('toJSONAsync', () => {
    it('supports Headers', async () => {
      const result = await toJSONAsync(Promise.resolve(EXAMPLE), {
        plugins: [HeadersPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<typeof EXAMPLE>>(result, {
        plugins: [HeadersPlugin],
      });
      expect(back).toBeInstanceOf(Headers);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('crossSerialize', () => {
    it('supports Headers', () => {
      const result = crossSerialize(EXAMPLE, {
        plugins: [HeadersPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports Headers', () => {
        const result = crossSerialize(EXAMPLE, {
          plugins: [HeadersPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports Headers', async () => {
      const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [HeadersPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports Headers', async () => {
        const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
          plugins: [HeadersPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports Headers', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(EXAMPLE), {
          plugins: [HeadersPlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports Headers', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(Promise.resolve(EXAMPLE), {
            plugins: [HeadersPlugin],
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports Headers', () => {
      const result = toCrossJSON(EXAMPLE, {
        plugins: [HeadersPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        plugins: [HeadersPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Headers);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports Headers', async () => {
      const example = Promise.resolve(EXAMPLE);
      const result = await toCrossJSONAsync(example, {
        plugins: [HeadersPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromCrossJSON<typeof example>(result, {
        plugins: [HeadersPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Headers);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports Headers', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(Promise.resolve(EXAMPLE), {
          plugins: [HeadersPlugin],
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
