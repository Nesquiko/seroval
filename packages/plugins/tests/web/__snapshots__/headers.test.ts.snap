// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Headers > crossSerialize > scoped > supports Headers 1`] = `"($R=>$R[0]=new Headers($R[1]=[$R[2]=["content-encoding","gzip"],$R[3]=["content-type","text/plain"]]))($R["example"])"`;

exports[`Headers > crossSerialize > supports Headers 1`] = `"$R[0]=new Headers($R[1]=[$R[2]=["content-encoding","gzip"],$R[3]=["content-type","text/plain"]])"`;

exports[`Headers > crossSerializeAsync > scoped > supports Headers 1`] = `"($R=>$R[0]=Promise.resolve($R[1]=new Headers($R[2]=[$R[3]=["content-encoding","gzip"],$R[4]=["content-type","text/plain"]])))($R["example"])"`;

exports[`Headers > crossSerializeAsync > supports Headers 1`] = `"$R[0]=Promise.resolve($R[1]=new Headers($R[2]=[$R[3]=["content-encoding","gzip"],$R[4]=["content-type","text/plain"]]))"`;

exports[`Headers > crossSerializeStream > scoped > supports Headers 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R["example"])"`;

exports[`Headers > crossSerializeStream > scoped > supports Headers 2`] = `"($R=>($R[6]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new Headers($R[3]=[$R[4]=["content-encoding","gzip"],$R[5]=["content-type","text/plain"]])))($R["example"])"`;

exports[`Headers > crossSerializeStream > supports Headers 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`Headers > crossSerializeStream > supports Headers 2`] = `"($R[6]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new Headers($R[3]=[$R[4]=["content-encoding","gzip"],$R[5]=["content-type","text/plain"]]))"`;

exports[`Headers > serialize > supports Headers 1`] = `"new Headers([["content-encoding","gzip"],["content-type","text/plain"]])"`;

exports[`Headers > serializeAsync > supports Headers 1`] = `"Promise.resolve(new Headers([["content-encoding","gzip"],["content-type","text/plain"]]))"`;

exports[`Headers > toCrossJSON > supports Headers 1`] = `"{"t":40,"i":0,"s":{"t":9,"i":1,"l":2,"a":[{"t":9,"i":2,"l":2,"a":[{"t":1,"s":"content-encoding"},{"t":1,"s":"gzip"}],"o":0},{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"}"`;

exports[`Headers > toCrossJSONAsync > supports Headers 1`] = `"{"t":12,"i":0,"s":1,"f":{"t":40,"i":1,"s":{"t":9,"i":2,"l":2,"a":[{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"content-encoding"},{"t":1,"s":"gzip"}],"o":0},{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"}}"`;

exports[`Headers > toCrossJSONStream > supports Headers 1`] = `"{"t":28,"i":0,"f":{"t":41,"i":1,"s":2}}"`;

exports[`Headers > toCrossJSONStream > supports Headers 2`] = `"{"t":29,"i":0,"a":[{"t":41,"i":6,"s":3},{"t":40,"i":2,"s":{"t":9,"i":3,"l":2,"a":[{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"content-encoding"},{"t":1,"s":"gzip"}],"o":0},{"t":9,"i":5,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"}]}"`;

exports[`Headers > toJSON > supports Headers 1`] = `"{"t":{"t":40,"i":0,"s":{"t":9,"i":1,"l":2,"a":[{"t":9,"i":2,"l":2,"a":[{"t":1,"s":"content-encoding"},{"t":1,"s":"gzip"}],"o":0},{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},"f":16383,"m":[]}"`;

exports[`Headers > toJSONAsync > supports Headers 1`] = `"{"t":{"t":12,"i":0,"s":1,"f":{"t":40,"i":1,"s":{"t":9,"i":2,"l":2,"a":[{"t":9,"i":3,"l":2,"a":[{"t":1,"s":"content-encoding"},{"t":1,"s":"gzip"}],"o":0},{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"}},"f":16383,"m":[]}"`;
